docs/src/pages/presets/writing-presets/index.md

```diff
@@ -0,0 +1,120 @@
+---
+
+[Storybook presets](../introduction/) are grouped collections of `babel`, `webpack`, and `addons` configurations that support specific use cases in Storybook, such as typescript or MDX support.
+
+This doc covers the [presets API](#presets-api) and how to use the presets mechanism for [advanced configuration](#advanced-configuration).
+
+## Presets API
+
+A preset is a set of hooks that is called by Storybook on initialization and can override configurations for `babel`, `webpack`, `addons`, and `entries`.
+
+Each configuration has a similar signature, accepting a base configuration object and options, as in this webpack example:
+
+```js
+export async function webpack(baseConfig, options) {
+  // modify or replace config
+  config.module.rules.push(/* some loader */);
+  return config;
+}
+```
+
+### Babel
+
+The babel functions `babel`, `babelDefault`, and `managerBabel` all configure babel in different ways.
+
+All functions take a [Babel configuration object](https://babeljs.io/docs/en/configuration) as their argument and can modify it or return a new object.
+
+For example, Storybook's Vue support uses presets internally and here's how it configures babel:
+
+```js
+export function babelDefault(config) {
+  return {
+    ...config,
+    presets: [...config.presets, require.resolve('babel-preset-vue')],
+  };
+}
+```
+
+- `babel` is applied to the preview config, after it has been initialized by storybook
+- `babelDefault` is applied to the preview config before any user presets have been applied
+- `managerBabel` is applied to the manager.
+
+### Webpack
+
+The webpack functions `webpack`, `webpackFinal`, and `managerWebpack` configure webpack.
+
+All functions take a [webpack4 configuration object](https://webpack.js.org/configuration/).
+
+For example, here is how Storybook automatically adopts `create-react-app`'s configuration if it's installed, where `applyCRAWebpackConfig` is a set of smart heuristics for modifying the input config.
+
+```js
+export function webpackFinal(config, { configDir }) {
+  if (!isReactScriptsInstalled()) {
+    logger.info('=> Using base config because react-scripts is not installed.');
+    return config;
+  }
+
+  logger.info('=> Loading create-react-app config.');
+  return applyCRAWebpackConfig(config, configDir);
+}
+```
+
+- `webpack` is applied to the preview config after it has been initialized by storybook
+- `webpackFinal` is applied to the preview config after all user presets have been applied
+- `webpackManager` is applied to the manager config
+
+### Addons
+
+The addon config `addons` allows you to add addons to Storybook from within a preset. For addons that require custom webpack/babel configuration, it is easier to simply install the preset, and it will take care of everything.
+
+For example, the Storysource preset contains the following code:
+
+```js
+export function addons(entry = []) {
+  return [...entry, require.resolve('@storybook/addon-storysource/register')];
```

Comment:
If I am using the knobs addon here in the exported addons function of the preset I am making, do I still need to use the `addDecorator(withKnobs)` in my config.js file? The docs are a little unclear about that in my opinion.

To add onto that thought, should a preset that utilizes the knob addon expose the things the knob addon does? (like `boolean`, `text`, etc?) Otherwise the consumer of said preset will also need to install @storybook/addon-knobs to get access to them. (is that what is envisioned)

Reply:
IMO the addons should export their own presets. That's what I'm doing with `addon-docs`. So you install the addon and then use its preset. Have not discussed this with @igor-dv regarding storysource.

Reply:
Okay. I'll make an issue and ask igor about what the expected usage/how to do preset of presets.